File: zeroday-server/src/main/kotlin/com/zeroday/model/Challenge.kt
package com.zeroday.model

import kotlinx.serialization.Serializable

/**
 * Static definition of a daily/weekly challenge. Challenges have a
 * cadence (`Daily` or `Weekly`) and a target metric the player has to
 * hit before they expire.
 */
@Serializable
data class ChallengeDefinition(
    val id: String,
    val name: String,
    val description: String,
    val cadence: ChallengeCadence,
    val event: AchievementEvent,
    val targetValue: Long,
    val rewardCredits: Long = 0L,
    val rewardXp: Long = 0L,
    val rewardReputation: Int = 0
)

enum class ChallengeCadence {
    DAILY,
    WEEKLY
}

/**
 * The player's active instance of a challenge. The server creates a new
 * list of these on each rotation (24h for daily, 7d for weekly); the
 * client just renders them.
 */
@Serializable
data class ActiveChallenge(
    val challengeId: String,
    val assignedAt: Long,
    val expiresAt: Long,
    var currentValue: Long = 0L,
    var completed: Boolean = false,
    var completedAt: Long? = null,
    var claimed: Boolean = false
) {
    fun percent(target: Long): Int =
        if (target == 0L) 100
        else ((currentValue.toDouble() / target) * 100).toInt().coerceIn(0, 100)
}

File: zero-server-csharp/src/ZeroDayMMO.Shared/Models/Challenge.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ZeroDayMMO.Shared.Models;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ChallengeCadence
{
    [JsonPropertyName("daily")]
    DAILY,
    [JsonPropertyName("weekly")]
    WEEKLY
}

public class ChallengeDefinition
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = "";

    [JsonPropertyName("name")]
    public string Name { get; set; } = "";

    [JsonPropertyName("description")]
    public string Description { get; set; } = "";

    [JsonPropertyName("cadence")]
    public ChallengeCadence Cadence { get; set; }

    [JsonPropertyName("event")]
    public AchievementEvent Event { get; set; }

    [JsonPropertyName("target_value")]
    public long TargetValue { get; set; }

    [JsonPropertyName("reward_credits")]
    public long RewardCredits { get; set; } = 0;

    [JsonPropertyName("reward_xp")]
    public long RewardXp { get; set; } = 0;

    [JsonPropertyName("reward_reputation")]
    public int RewardReputation { get; set; } = 0;
}

public class ActiveChallenge
{
    [JsonPropertyName("challenge_id")]
    public string ChallengeId { get; set; } = "";

    [JsonPropertyName("assigned_at")]
    public long AssignedAt { get; set; }

    [JsonPropertyName("expires_at")]
    public long ExpiresAt { get; set; }

    [JsonPropertyName("current_value")]
    public long CurrentValue { get; set; } = 0;

    [JsonPropertyName("completed")]
    public bool Completed { get; set; } = false;

    [JsonPropertyName("completed_at")]
    public long? CompletedAt { get; set; }

    [JsonPropertyName("claimed")]
    public bool Claimed { get; set; } = false;

    public int Percent(long target)
    {
        if (target == 0) return 100;
        return Math.Clamp((int)((double)CurrentValue / target * 100), 0, 100);
    }
}
